### DISCUSSION TITLE
Custom board?

### DISCUSSION THREAD

**Discussion by glosair** [raomin/ESPAltherma#144] [Ideas]
Would anyone be interested in an opensource custom board If I design one? 

Things so far.

Switching power supply.
Serial voltage shifters. (5V to 3.3V).

**Comment by raomin** [comment]
Hello @glosair, thanks for the proposal.
I don't think it's needed, or at least I don't personally feel the need for that.
ESPAltherma was designed to be plug and play. ESP32 are 5v tolerant, Daikin ttl is 3v tolerant, and most of the time, you can steal power from the heatpump itself...

Now, if others are interested... I'll leave the issue open for a while.

**Comment by andreapier** [comment]
I'd be interested, but only if the board also includes relays (ideally 4: 1x heating, 1x cooling, 2x Smart Grid). Only 2 for smart grid would still be cool, though.

**Comment by glosair** [comment]
ESP32's aren't 5V tolerant, you might get away with it but.....

This board got extremely hot when connected as per the instructions.  When I installed level shifters and connected I also found my version K board would not power the ESP32 pictured via it's LDO regulator. 

![image](https://user-images.githubusercontent.com/58359462/189913378-609d37c9-e4fa-4fc3-a4c2-09008a181294.png)

![IMG_20220913_144235](https://user-images.githubusercontent.com/58359462/189934552-b162d36e-801f-490e-a843-b278a4dc11cc.jpg)

**Comment by mauromorello** [comment]
I used a LOLIN32 where there is a BMS onboard and can take from 3 to 7 volt in input; I put Vin  from altherma in the battery connector; 

From a forum found on web: An alternative is to power at 4.2V via the battery connector. The TP4054 can take up to 7V on its BAT pin, and the charger won't enable unless the voltage drops below 4.2V, so one could also supply, say, 5V via this connector. The D2 diode will protect USB.

![image](https://user-images.githubusercontent.com/3259982/189937090-720aae40-e428-4004-ae2e-8729155103f7.png)

**Comment by glosair** [comment]
> I'd be interested, but only if the board also includes relays (ideally 4: 1x heating, 1x cooling, 2x Smart Grid). Only 2 for smart grid would still be cool, though.

No problems there, will have a look for suitable surface mount relays, will they need to be NO, NC or Both? 

> I used a LOLIN32 where there is a BMS onboard and can take from 3 to 7 volt in input; I put Vin from altherma in the battery connector;
> 
> From a forum found on web: An alternative is to power at 4.2V via the battery connector. The TP4054 can take up to 7V on its BAT pin, and the charger won't enable unless the voltage drops below 4.2V, so one could also supply, say, 5V via this connector. The D2 diode will protect USB.
> 
I think it might be a problem with the LDO (analogue) regulator as the 5V on the serial interface will get clamped down to the 3.3V rail by the anti static protection and feed the chip with 3.3V + 0.6V (diode drop) ==> 3.9V and any serial coms will create a very unstable supply voltage.  It looks like the boards that don't have a problem use a switching regulator as if running from battery you don't want to dump energy as heat from a regulator.

**Comment by andreapier** [comment]
> No problems there, will have a look for suitable surface mount relays, will they need to be NO, NC or Both?

NO would work best. ESP Altherma supports both through a configuration switch, tough.

**Comment by alexdelli** [comment]
with relays included yes

**Comment by raomin** [comment]
Again, I don't want to oppose or anything. Building a board is a fun project, go for it!
But, also, consider having a look at [existing esp32 boards with relays](https://www.aliexpress.com/wholesale?SearchText=esp32+relay+board).
ESP32 are 5V (open) tolerant for IO pins, you should have no current dissipation (nor extra heat).
A good addition to these (existing? future?) boards would be proper connection to X10A. but then the cable length can greatly vary from installation to installation...

_I'll move this issue to the newly open discussion board_

**Reply by DieterB61** [reply]
Hi raomin, 

my ESP32 development board is now running with ESPAltherma and is connected to the 5V of the X10A of my Altherma 3. As you said here and elsewhere, there is no problem using that 5V (I don't have relays up to now).

My question: can we also use 5V relay modules connected to the GOIO pins of ESP32 or must that be 3.3V relays? And when I use additional relays (e.g. for the use with X85A), ist the power coming from X10A sufficient also for that?

Regarding your link to boards with relays: at the moment I mainly found ESP8266 combined with relays. Is ESP8266 supported by ESPAltherma? 

If ESPAltherma works with ESP8266 one could also take a Sonoff 4ch pro. There is direct access to TR,RX and thus it can be flashed (see Tasmota). The Sonoff 4ch pro includes a power source, a housing, 4 relays and additional buttons that could be used to control energy/SG directly at the HP (if somebody add the code :-) ).

**Reply by glosair** [reply]
The relay boards normally have the relays optoisolated so would require 5v for the relay drive power but would accept 3.3V on the inputs. You can get better boards like this that are way more flexible. 
![IMG_20230419_110404](https://user-images.githubusercontent.com/58359462/233042793-7c4f0376-94e2-4cb0-bdc9-f5a9a8ab7b29.jpg)

**Comment by DieterB61** [comment]
You mean better than the Sonoff (no buttons, no housing)? 
Which processor does your board have?

**Reply by glosair** [reply]
The one pictures has an ESP32, versions are available based on esp8266, but remember you will only be able to use the floating contacts of the Sonoff as the rest of the device is almost certainly floating mains voltage so if you tried to connect it to read data from the Altherma  you might be buying a new board for your heat pump!

**Reply by DieterB61** [reply]
Many thanks for that information, I think I will not try that :-)

**Reply by DieterB61** [reply]
> The one pictures has an ESP32, versions are available based on esp8266, ...

I ordered the one with ESP-32 but got one with ESP-12F :-(
It is worth to give that one a try?

**Comment by bashers222** [comment]
I'm using a ESP8266 with my unit as i have so many available.
I did have to tinker in the platform.ini to get it to work though, but its not too hard

However, as to power usage **beware** connecting up too many devices.
I currently have EPSaltherma unit, plus another ESP8266 for ESPHome with 2 relays to trigger 2 zone relays (due to the ability to hide it all in the Daikin indoor enclosure powered from the Daikin board)
What i've seen twice is the Daikin controller board hanging, requiring reboot. I'm now powering the second unit off its own PSU

**Reply by maromme** [reply]
edit: didn't read the full thread... similar answers above 

Well if you consider a all in one Board why don't you use something like that:
https://www.ebay.de/itm/274712888599

some more information on that board:
https://werner.rothschopf.net/microcontroller/202108_esp8266_esp12f_relay_x8_en.htm

You can find it with different number of relays.

I get the one with 8 relays next week for another project, but wouldn't mind testing it with the espaltherma. The 8 really board is fine for heat/cool on/off, SG1, SG2, and 4x power demand shaping. Maybe I stick to it if the test is fine 🤔

**Reply by DieterB61** [reply]
I already have the one with 4 relays and will also try it soon.
I'm confused: is there already a ESPAltherma code that allows to switch the 4 or 8 relays via MQTT variables?

**Reply by maromme** [reply]
no, the main branche is not supporting more than heat on/off and SG1 and SG2. But development is going on and with a little bit of coding it is possible. bit guess soon there is more coming

**Reply by maromme** [reply]
I got it working. But it's not completely straight forward.

To upload the first time you've got to connect pin 0 to ground (and power cycle). Then you need the latest sourcecode by raomin. He fixed a small issue yesterday. Thanks for that!

Pin configuration as follows. But be aware that the third relay (pin 0) is switched on. Guess pin 0 is standard high in the esp if not used and thereby defined.

```
//Default GPIO PINs for Serial2:
#define RX_PIN    1// Pin connected to the TX pin of X10A 
#define TX_PIN    3// Pin connected to the RX pin of X10A
#endif

#define PIN_THERM 4// Pin connected t   o the thermostat relay (normally open)

//Smart grid control - Optional:
//Uncomment and set to enable SG mqtt functions
#define PIN_SG1 12// Pin connected to dry contact SG 1 relay (normally open)
#define PIN_SG2 13// Pin connected to dry contact SG 2 relay (normally open)
// Define if your SG relay board is Low or High triggered (signal pins)
// Only uncomment one of them
#define SG_RELAY_HIGH_TRIGGER
//#define SG_RELAY_LOW_TRIGGER
```

Another thing: the on/off (PIN_THERM) is inverted. If I switch in homeassistant on, the relay is off and the other way around. You could connect the contact of the heatpump the other way around, but as in our case the heating is mostly off it's no point in having the relay using power to be "on"
